helix-term/src/ui/picker.rs

```diff
@@ -283,43 +292,57 @@ impl<T: Item + 'static> Component for FilePicker<T> {
                 })
                 .unwrap_or(0);
 
-            let offset = Position::new(first_line, 0);
+            let offset = ViewPosition {
+                anchor: doc.text().line_to_char(first_line),
+                horizontal_offset: 0,
+                vertical_offset: 0,
+            };
 
-            let mut highlights =
-                EditorView::doc_syntax_highlights(doc, offset, area.height, &cx.editor.theme);
+            let mut highlights = EditorView::doc_syntax_highlights(
+                doc,
+                offset.anchor,
+                area.height,
+                &cx.editor.theme,
+            );
             for spans in EditorView::doc_diagnostics_highlights(doc, &cx.editor.theme) {
                 if spans.is_empty() {
                     continue;
                 }
                 highlights = Box::new(helix_core::syntax::merge(highlights, spans));
             }
-            EditorView::render_text_highlights(
+            let mut decorations: Vec<Box<dyn LineDecoration>> = Vec::new();
```

Comment:
Do we need to know all the decorations ahead of time? Do we only fetch the decorations for the visible line range?

Reply:
LineDecorations are a callback that is called for each visual line and can draw that line. So only computing the line decorations for the visual range is up to the specific implementation (bests done before construction). The callback is only called for the visual lines that are actually rendered tough so even just a naive implementation could check whether we are at the right line with a simple if.

Reply:
The following decorator that I implemented for the DAP highlight is a good example because it's quite simple:
``` rust
 let line_decoration = move |renderer: &mut TextRenderer, pos: LinePos| {
    if pos.doc_line != line {
        return;
    }
    renderer
        .surface
        .set_style(Rect::new(area.x, pos.visual_line, area.width, 1), style);
};

line_decorations.push(Box::new(line_decoration));
```

Right now I only check inside the callback whether we are in the right line. Adding a check whether `line` is between the first line and the last line would be trivial and only calling `line_decorations.push(Box::new(line_decoration));` is the line is inbounds would be trivial.

So the `LineDecorator` is really just supposed to be a very thin abstraction that aims to be as flexible as possible and leave all these details up to the caller.